src/ray/gcs/gcs_server/gcs_placement_group_manager.cc

```diff
@@ -164,29 +164,32 @@ void GcsPlacementGroupManager::RegisterPlacementGroup(
     }
   }
 
-  // Mark the callback as pending and invoke it after the placement_group has been
-  // successfully created.
-  placement_group_to_register_callback_[placement_group->GetPlacementGroupID()] =
-      std::move(callback);
+  placement_group_to_register_callbacks_[placement_group->GetPlacementGroupID()]
+      .emplace_back(std::move(callback));
   registered_placement_groups_.emplace(placement_group->GetPlacementGroupID(),
                                        placement_group);
   pending_placement_groups_.emplace_back(placement_group);
 
   RAY_CHECK_OK(gcs_table_storage_->PlacementGroupTable().Put(
       placement_group_id, placement_group->GetPlacementGroupTableData(),
       [this, placement_group_id, placement_group](Status status) {
+        // The backend storage is supposed to be reliable, so the status must be ok.
         RAY_CHECK_OK(status);
-        if (!registered_placement_groups_.contains(placement_group_id)) {
-          auto iter = placement_group_to_register_callback_.find(placement_group_id);
-          if (iter != placement_group_to_register_callback_.end()) {
-            std::stringstream stream;
-            stream << "Placement group of id " << placement_group_id
-                   << " has been removed before registration.";
-            iter->second(Status::NotFound(stream.str()));
-            placement_group_to_register_callback_.erase(iter);
+        if (registered_placement_groups_.contains(placement_group_id)) {
+          auto iter = placement_group_to_register_callbacks_.find(placement_group_id);
+          auto callbacks = std::move(iter->second);
+          placement_group_to_register_callbacks_.erase(iter);
+          for (const auto &callback : callbacks) {
+            callback(status);
           }
-        } else {
           SchedulePendingPlacementGroups();
+        } else {
+          // The placement group registration is synchronous, so if we found the placement
```

Comment:
After discussed with @ffbin ,  we don't need to invoke `RAY_CHECK(placement_group_to_register_callback_.count(placement_group_id) == 0)` here,  because we remove the placement group from registered_placement_groups_ and placement_group_to_register_callbacks_  only in the `RemovePlacementGroup`, and this will also introduced the risk of GCS server crash if we do this.  @rkooo567

Reply:
Hmm can you at least make the log message to ERROR and say "This is a bug that should be addressed. Please report to Ray Github issue if you see this message"?

Reply:
I still prefer to have a RAY_CHECK for strong consistency. What's the reason you guys prefer the log message? If we see the check failure, that means we just need to fix the issue rather than ignoring with the log messages.